amethyst_rendy/src/bundle.rs

```diff
@@ -0,0 +1,621 @@
+        }
+
+        for dep in deps {
+            if let Some(node) = ctx.get_pass_node(dep) {
+                subpass.add_dependency(node);
+            }
+        }
+
+        pass.add_subpass(subpass);
+        ctx.submit_pass(self.key, pass)?;
+        Ok(())
+    }
+}
+
+/// An action that represents a single transformation to the
+/// render graph, e.g. addition of single render group.
+///
+/// TODO: more actions needed for e.g. splitting pass into subpasses.
+#[derive(Debug)]
+pub enum RenedrableAction<B: Backend> {
+    /// Register single render group for evaluation during target rendering
+    RenderGroup(Box<dyn RenderGroupBuilder<B, Resources>>),
+}
+
+impl<B: Backend> RenedrableAction<B> {
+    fn colors(&self) -> usize {
+        match self {
+            RenedrableAction::RenderGroup(g) => g.colors(),
+        }
+    }
+
+    fn depth(&self) -> bool {
+        match self {
+            RenedrableAction::RenderGroup(g) => g.depth(),
+        }
+    }
+}
+
+/// Trait for easy conversion of various types into `RenedrableAction` shell.
+pub trait IntoAction<B: Backend> {
+    /// Convert to `RenedrableAction`.
+    fn into(self) -> RenedrableAction<B>;
+}
+
+impl<B: Backend, G: RenderGroupBuilder<B, Resources> + 'static> IntoAction<B> for G {
+    fn into(self) -> RenedrableAction<B> {
+        RenedrableAction::RenderGroup(Box::new(self))
+    }
+}
+
+/// Collection of predefined constants for action ordering in the targets.
+/// Two actions with the same order will be applied in their insertion order.
+/// The list is provided mostly as a comparison point. If you can't find the exact
+/// ordering you need, provide custom `i32` that fits into the right place.
+///
+/// Crates that provide custom render plugins using their own orders should export
+/// similar enum with ordering they have added.
+#[derive(Debug)]
+#[repr(i32)]
+pub enum RenderOrder {
+    /// register before all opaques
+    BeforeOpaque = 90,
+    /// register for rendering opaque objects
+    Opaque = 100,
+    /// register after rendering opaque objects
+    AfterOpaque = 110,
+    /// register before rendering transparent objects
+    BeforeTransparent = 190,
+    /// register for rendering transparent objects
+    Transparent = 200,
+    /// register after rendering transparent objects
+    AfterTransparent = 210,
+    /// register as post effect in linear color space
+    LinearPostEffects = 300,
+    /// register as tonemapping step
+    ToneMap = 400,
+    /// register as post effect in display color space
+    DisplayPostEffects = 500,
+}
+
+impl Into<i32> for RenderOrder {
+    fn into(self) -> i32 {
+        self as i32
+    }
+}
+
+/// An identifier for render target used in render plugins.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Target {
+    /// Default render target for most operations.
+    /// Usually the one that gets presented to the window.
+    Main,
+    /// Custom render target identifier.
+    Custom(&'static str),
```

Comment:
It would poison a lot. You would have to surface that generic up to the rendering bundle. I see no point. This is not serializable or anything as well, so no problems with `&'static` here. The reason i wanted strings is that collision is not very likely and at the same time it doesn't require typesystem magic. We use enums for all the intertnal stuff, can as well make a few extra if something is common (like I just did for shadows). Everything user-defined must be hard to collide, easy to use and not pollute the API.

Reply:
sounds good :thumbsup: